client/commands.c

```diff
@@ -48,21 +58,189 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
+	{ "pool"       , xdag_com_pool},
+	{ "run"        , xdag_com_run},
+	{ "state"      , xdag_com_state},
+	{ "stats"      , xdag_com_stats},
+	{ "terminate"  , xdag_com_terminate},
+	{ "xfer"       ,(xdag_com_func_t)NULL},
+	{ "help"       , xdag_com_help},
+	{ (char *)NULL ,(xdag_com_func_t)NULL}
+};
+
+int xdag_com_account(char* args, FILE* out) {
+	processAccountCommand(args, out);
+	return 0;
+}
+
+int xdag_com_balance(char * args, FILE* out) {
+	processBalanceCommand(args, out);
+	return 0;
+}
+
+int xdag_com_block(char * args, FILE* out) {
+	processBlockCommand(args, out);
+	return 0;
+}
+
+int xdag_com_lastblocks(char * args, FILE* out) {
+	processLastBlocksCommand(args, out);
+	return 0;
+}
+
+int xdag_com_keyGen(char * args, FILE* out) {
+	processKeyGenCommand(out);
+	return 0;
+}
+
+int xdag_com_level(char * args, FILE* out) {
+	processLevelCommand(args, out);
+	return 0;
+}
+
+int xdag_com_mining(char * args, FILE* out) {
+	processMiningCommand(args, out);
+	return 0;
+}
+
+int xdag_com_net(char * args, FILE* out) {
+	processNetCommand(args, out);
+	return 0;
+}
+
+int xdag_com_pool(char * args, FILE* out) {
+	processPoolCommand(args, out);
+	return 0;
+}
+
+int xdag_com_miners(char * args, FILE* out) {
+	processMinersCommand(args, out);
+	return 0;
+}
+
+int xdag_com_stats(char * args, FILE* out) {
+	processStatsCommand(out);
+	return 0;
+}
+
+int xdag_com_state(char * args, FILE* out) {
+	fprintf(out, "%s\n", get_state());
+	return 0;
+}
+
+int xdag_com_run(char * args, FILE* out) {
+	g_xdag_run = 1;
+	return 0;
+}
+
+int xdag_com_terminate(char * args, FILE* out) {
+	processExitCommand();
+	return -1;
+}
+
+int xdag_com_exit(char * args, FILE* out) {
+	processExitCommand();
+	return -1;
+}
+
+int xdag_com_help(char *args, FILE* out) {
+	printHelp(out);
+	return 0;
+}
+
+char ** xdag_com_completion(const char *text, int start, int end) {
+	char **matches = (char **)NULL;;
+	if (start == 0)
+        	matches = rl_completion_matches(text, xdag_com_generator);
+	return (matches);
+}
+
+char* xdag_com_generator(const char* text, int state) {
+	static int list_index, len;
+	char *name;
+	if (!state) {
+        	list_index = 0;
+        	len = strlen(text);
+	 }
+    
+	while ((name = commands[list_index].name)) {
+		list_index++;
+		if (strncmp(name, text, len) == 0)
+			return (strdup(name));
+	}
+	return ((char *)NULL);
+}
+
+XDAG_COMMAND* find_xdag_command(char *name) {
+	int i;
+	for (i = 0; commands[i].name; i++)
+	if (strcmp(name, commands[i].name) == 0)
+		return (&commands[i]);
+	return ((XDAG_COMMAND *)NULL);
+}
+
 void startCommandProcessing(int transportFlags)
 {
-	char cmd[XDAG_COMMAND_MAX] = {0};
-
+	char cmd[XDAG_COMMAND_MAX];
 	if(!(transportFlags & XDAG_DAEMON)) printf("Type command, help for example.\n");
-	for(;;) {
-		if(transportFlags & XDAG_DAEMON) sleep(100);
-		else {
+    
+#if !defined(_WIN32) && !defined(_WIN64)
+	rl_readline_name = "xdag";
+	rl_attempted_completion_function = xdag_com_completion;
```

Comment:
New memory malloced in xdag_com_completion, how is it handled? Freed by readline lib or manually?

Reply:
xdag_com_completion malloc memory will free in readline lib auto:

in readline lib version 6.3 , complete.c:

1160   if (rl_attempted_completion_function)
1161     {
1162       matches = (*rl_attempted_completion_function) (text, start, end);
1163       if (RL_SIG_RECEIVED())
1164         {
1165           _rl_free_match_list (matches);
1166           matches = 0;
1167           RL_CHECK_SIGNALS ();
1168         }
1169 
1170       if (matches || rl_attempted_completion_over)
1171         {
1172           rl_attempted_completion_over = 0;
1173           return (matches);
1174         }
1175     }

_rl_free_match_list will free memory:

1897 void
1898 _rl_free_match_list (matches)
1899      char **matches;
1900 {
1901   register int i;
1902 
1903   if (matches == 0)
1904     return;
1905 
1906   for (i = 0; matches[i]; i++)
1907     xfree (matches[i]);
1908   xfree (matches);
1909 }